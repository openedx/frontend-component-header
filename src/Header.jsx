import React, { useContext } from 'react';
import Responsive from 'react-responsive';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { AppContext } from '@edx/frontend-platform/react';
import {
  APP_CONFIG_INITIALIZED,
  ensureConfig,
  mergeConfig,
  getConfig,
  subscribe,
} from '@edx/frontend-platform';

import DesktopHeader from './DesktopHeader';
import MobileHeader from './MobileHeader';

import messages from './Header.messages';

ensureConfig([
  'LMS_BASE_URL',
  'LOGOUT_URL',
  'LOGIN_URL',
  'SITE_NAME',
  'LOGO_URL',
], 'Header component');

subscribe(APP_CONFIG_INITIALIZED, () => {
  mergeConfig({
    LOGISTRATION_MINIMAL_HEADER: !!process.env.LOGISTRATION_MINIMAL_HEADER,
  }, 'Header additional config');
});

function Header({ intl }) {
  const { authenticatedUser, config } = useContext(AppContext);

  const mainMenu = [
    {
      type: 'item',
      href: `${config.LMS_BASE_URL}/dashboard`,
      content: intl.formatMessage(messages['header.links.courses']),
    },
  ];

  const userMenu = authenticatedUser === null ? [] : [
    {
      type: 'item',
      href: `${config.LMS_BASE_URL}/dashboard`,
      content: intl.formatMessage(messages['header.user.menu.dashboard']),
    },
    {
      type: 'item',
      href: `${config.LMS_BASE_URL}/u/${authenticatedUser.username}`,
      content: intl.formatMessage(messages['header.user.menu.profile']),
    },
    {
      type: 'item',
      href: `${config.LMS_BASE_URL}/account/settings`,
      content: intl.formatMessage(messages['header.user.menu.account.settings']),
    },
    {
      type: 'item',
      href: config.LOGOUT_URL,
      content: intl.formatMessage(messages['header.user.menu.logout']),
    },
  ];

  const loggedOutItems = [
    {
      type: 'item',
      href: config.LOGIN_URL,
      content: intl.formatMessage(messages['header.user.menu.login']),
    },
    {
      type: 'item',
      href: `${config.LMS_BASE_URL}/register`,
      content: intl.formatMessage(messages['header.user.menu.register']),
    },
  ];

  const props = {
    logo: config.LOGO_URL,
    logoAltText: config.SITE_NAME,
    siteName: config.SITE_NAME,
    logoDestination: `${config.LMS_BASE_URL}/dashboard`,
    loggedIn: authenticatedUser !== null,
    username: authenticatedUser !== null ? authenticatedUser.username : null,
    avatar: authenticatedUser !== null ? authenticatedUser.avatar : null,
    mainMenu: getConfig().LOGISTRATION_MINIMAL_HEADER ? [] : mainMenu,
    userMenu,
    loggedOutItems: getConfig().LOGISTRATION_MINIMAL_HEADER ? [] : loggedOutItems,
  };

  return (
    <>
      <Responsive maxWidth={768}>
        <MobileHeader {...props} />
      </Responsive>
      <Responsive minWidth={769}>
        <DesktopHeader {...props} />
      </Responsive>
    </>
  );
}

Header.propTypes = {
  intl: intlShape.isRequired,
};

export default injectIntl(Header);
