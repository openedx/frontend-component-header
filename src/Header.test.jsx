import React from 'react';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import TestRenderer from 'react-test-renderer';
import { AppContext } from '@edx/frontend-platform/react';
import { Context as ResponsiveContext } from 'react-responsive';

import Header from './index';

describe('<Header />', () => {
  it('renders correctly for anonymous desktop', () => {
    const component = (
      <ResponsiveContext.Provider value={{ width: 1280 }}>
        <IntlProvider locale="en" messages={{}}>
          <AppContext.Provider
            value={{
              authenticatedUser: null,
              config: {
                LMS_BASE_URL: process.env.LMS_BASE_URL,
                SITE_NAME: process.env.SITE_NAME,
                LOGIN_URL: process.env.LOGIN_URL,
                LOGOUT_URL: process.env.LOGOUT_URL,
                LOGO_URL: process.env.LOGO_URL,
              },
            }}
          >
            <Header />
          </AppContext.Provider>
        </IntlProvider>
      </ResponsiveContext.Provider>
    );

    const wrapper = TestRenderer.create(component);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('renders correctly for authenticated desktop', () => {
    const component = (
      <ResponsiveContext.Provider value={{ width: 1280 }}>
        <IntlProvider locale="en" messages={{}}>
          <AppContext.Provider
            value={{
              authenticatedUser: {
                userId: 'abc123',
                username: 'edX',
                roles: [],
                administrator: false,
              },
              config: {
                LMS_BASE_URL: process.env.LMS_BASE_URL,
                SITE_NAME: process.env.SITE_NAME,
                LOGIN_URL: process.env.LOGIN_URL,
                LOGOUT_URL: process.env.LOGOUT_URL,
                LOGO_URL: process.env.LOGO_URL,
              },
            }}
          >
            <Header />
          </AppContext.Provider>
        </IntlProvider>
      </ResponsiveContext.Provider>
    );

    const wrapper = TestRenderer.create(component);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('renders correctly for anonymous mobile', () => {
    const component = (
      <ResponsiveContext.Provider value={{ width: 500 }}>
        <IntlProvider locale="en" messages={{}}>
          <AppContext.Provider
            value={{
              authenticatedUser: null,
              config: {
                LMS_BASE_URL: process.env.LMS_BASE_URL,
                SITE_NAME: process.env.SITE_NAME,
                LOGIN_URL: process.env.LOGIN_URL,
                LOGOUT_URL: process.env.LOGOUT_URL,
                LOGO_URL: process.env.LOGO_URL,
              },
            }}
          >
            <Header />
          </AppContext.Provider>
        </IntlProvider>
      </ResponsiveContext.Provider>
    );

    const wrapper = TestRenderer.create(component);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('renders correctly for authenticated mobile', () => {
    const component = (
      <ResponsiveContext.Provider value={{ width: 500 }}>
        <IntlProvider locale="en" messages={{}}>
          <AppContext.Provider
            value={{
              authenticatedUser: {
                userId: 'abc123',
                username: 'edX',
                roles: [],
                administrator: false,
              },
              config: {
                LMS_BASE_URL: process.env.LMS_BASE_URL,
                SITE_NAME: process.env.SITE_NAME,
                LOGIN_URL: process.env.LOGIN_URL,
                LOGOUT_URL: process.env.LOGOUT_URL,
                LOGO_URL: process.env.LOGO_URL,
              },
            }}
          >
            <Header />
          </AppContext.Provider>
        </IntlProvider>
      </ResponsiveContext.Provider>
    );

    const wrapper = TestRenderer.create(component);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
