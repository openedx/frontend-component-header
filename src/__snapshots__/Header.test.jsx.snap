// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders correctly for anonymous desktop 1`] = `
<header
  className="site-header-desktop"
>
  <div
    className="container-fluid"
  >
    <div
      className="nav-container position-relative d-flex align-items-center"
    >
      <a
        className="logo"
        href="http://localhost:18000/dashboard"
      >
        <img
          alt="edX"
          className="d-block"
          src="https://edx-cdn.org/v3/default/logo.svg"
        />
      </a>
      <nav
        aria-label="Main"
        className="nav main-nav"
      >
        <a
          className="nav-link"
          href="http://localhost:18000/dashboard"
        >
          Courses
        </a>
      </nav>
      <nav
        aria-label="Secondary"
        className="nav secondary-menu-container align-items-center ml-auto"
      >
        <a
          className="btn mr-2 btn-link"
          href="http://localhost:18000/login"
        >
          Login
        </a>
        <a
          className="btn mr-2 btn-outline-primary"
          href="http://localhost:18000/register"
        >
          Sign Up
        </a>
      </nav>
    </div>
  </div>
</header>
`;

exports[`<Header /> renders correctly for anonymous mobile 1`] = `
<header
  aria-label="Main"
  className="site-header-mobile d-flex justify-content-between align-items-center shadow sticky-top"
>
  <div
    className="w-100 d-flex justify-content-start"
  >
    <div
      className="menu position-static"
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        aria-expanded={false}
        aria-haspopup="menu"
        aria-label="Main Menu"
        className="menu-trigger icon-button"
        onClick={[Function]}
        title="Main Menu"
      >
        <svg
          aria-hidden={true}
          focusable="false"
          height="24px"
          role="img"
          style={
            Object {
              "height": "1.5rem",
              "width": "1.5rem",
            }
          }
          version="1.1"
          viewBox="0 0 24 24"
          width="24px"
        >
          <rect
            fill="currentColor"
            height="2"
            width="20"
            x="2"
            y="5"
          />
          <rect
            fill="currentColor"
            height="2"
            width="20"
            x="2"
            y="11"
          />
          <rect
            fill="currentColor"
            height="2"
            width="20"
            x="2"
            y="17"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="w-100 d-flex justify-content-center"
  >
    <a
      className="logo"
      href="http://localhost:18000/dashboard"
      itemType="http://schema.org/Organization"
    >
      <img
        alt="edX"
        className="d-block"
        src="https://edx-cdn.org/v3/default/logo.svg"
      />
    </a>
  </div>
  <div
    className="w-100 d-flex justify-content-end align-items-center"
  >
    <nav
      aria-label="Secondary"
      className="menu position-static"
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        aria-expanded={false}
        aria-haspopup="menu"
        aria-label="Account Menu"
        className="menu-trigger icon-button"
        onClick={[Function]}
        title="Account Menu"
      >
        <span
          className="avatar overflow-hidden d-inline-flex rounded-circle null"
          style={
            Object {
              "height": "1.5rem",
              "width": "1.5rem",
            }
          }
        >
          <svg
            aria-hidden={true}
            focusable="false"
            height="24px"
            role="img"
            style={
              Object {
                "height": "1.5rem",
                "width": "1.5rem",
              }
            }
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
          >
            <path
              d="M4.10255106,18.1351061 C4.7170266,16.0581859 8.01891846,14.4720277 12,14.4720277 C15.9810815,14.4720277 19.2829734,16.0581859 19.8974489,18.1351061 C21.215206,16.4412566 22,14.3122775 22,12 C22,6.4771525 17.5228475,2 12,2 C6.4771525,2 2,6.4771525 2,12 C2,14.3122775 2.78479405,16.4412566 4.10255106,18.1351061 Z M12,24 C5.372583,24 0,18.627417 0,12 C0,5.372583 5.372583,0 12,0 C18.627417,0 24,5.372583 24,12 C24,18.627417 18.627417,24 12,24 Z M12,13 C9.790861,13 8,11.209139 8,9 C8,6.790861 9.790861,5 12,5 C14.209139,5 16,6.790861 16,9 C16,11.209139 14.209139,13 12,13 Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </nav>
  </div>
</header>
`;

exports[`<Header /> renders correctly for authenticated desktop 1`] = `
<header
  className="site-header-desktop"
>
  <div
    className="container-fluid"
  >
    <div
      className="nav-container position-relative d-flex align-items-center"
    >
      <a
        className="logo"
        href="http://localhost:18000/dashboard"
      >
        <img
          alt="edX"
          className="d-block"
          src="https://edx-cdn.org/v3/default/logo.svg"
        />
      </a>
      <nav
        aria-label="Main"
        className="nav main-nav"
      >
        <a
          className="nav-link"
          href="http://localhost:18000/dashboard"
        >
          Courses
        </a>
      </nav>
      <nav
        aria-label="Secondary"
        className="nav secondary-menu-container align-items-center ml-auto"
      >
        <div
          className="menu null"
          onKeyDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <button
            aria-expanded={false}
            aria-haspopup="menu"
            aria-label="Account menu for edX"
            className="menu-trigger btn btn-outline-primary d-inline-flex align-items-center pl-2 pr-3"
            onClick={[Function]}
          >
            <span
              className="avatar overflow-hidden d-inline-flex rounded-circle mr-2"
              style={
                Object {
                  "height": "1.5em",
                  "width": "1.5em",
                }
              }
            >
              <svg
                aria-hidden={true}
                focusable="false"
                height="24px"
                role="img"
                style={
                  Object {
                    "height": "1.5em",
                    "width": "1.5em",
                  }
                }
                version="1.1"
                viewBox="0 0 24 24"
                width="24px"
              >
                <path
                  d="M4.10255106,18.1351061 C4.7170266,16.0581859 8.01891846,14.4720277 12,14.4720277 C15.9810815,14.4720277 19.2829734,16.0581859 19.8974489,18.1351061 C21.215206,16.4412566 22,14.3122775 22,12 C22,6.4771525 17.5228475,2 12,2 C6.4771525,2 2,6.4771525 2,12 C2,14.3122775 2.78479405,16.4412566 4.10255106,18.1351061 Z M12,24 C5.372583,24 0,18.627417 0,12 C0,5.372583 5.372583,0 12,0 C18.627417,0 24,5.372583 24,12 C24,18.627417 18.627417,24 12,24 Z M12,13 C9.790861,13 8,11.209139 8,9 C8,6.790861 9.790861,5 12,5 C14.209139,5 16,6.790861 16,9 C16,11.209139 14.209139,13 12,13 Z"
                  fill="currentColor"
                />
              </svg>
            </span>
            edX
             
            <svg
              aria-hidden={true}
              focusable="false"
              height="16px"
              role="img"
              version="1.1"
              viewBox="0 0 16 16"
              width="16px"
            >
              <path
                d="M7,4 L7,8 L11,8 L11,10 L5,10 L5,4 L7,4 Z"
                fill="currentColor"
                transform="translate(8.000000, 7.000000) rotate(-45.000000) translate(-8.000000, -7.000000) "
              />
            </svg>
          </button>
        </div>
      </nav>
    </div>
  </div>
</header>
`;

exports[`<Header /> renders correctly for authenticated mobile 1`] = `
<header
  aria-label="Main"
  className="site-header-mobile d-flex justify-content-between align-items-center shadow sticky-top"
>
  <div
    className="w-100 d-flex justify-content-start"
  >
    <div
      className="menu position-static"
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        aria-expanded={false}
        aria-haspopup="menu"
        aria-label="Main Menu"
        className="menu-trigger icon-button"
        onClick={[Function]}
        title="Main Menu"
      >
        <svg
          aria-hidden={true}
          focusable="false"
          height="24px"
          role="img"
          style={
            Object {
              "height": "1.5rem",
              "width": "1.5rem",
            }
          }
          version="1.1"
          viewBox="0 0 24 24"
          width="24px"
        >
          <rect
            fill="currentColor"
            height="2"
            width="20"
            x="2"
            y="5"
          />
          <rect
            fill="currentColor"
            height="2"
            width="20"
            x="2"
            y="11"
          />
          <rect
            fill="currentColor"
            height="2"
            width="20"
            x="2"
            y="17"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="w-100 d-flex justify-content-center"
  >
    <a
      className="logo"
      href="http://localhost:18000/dashboard"
      itemType="http://schema.org/Organization"
    >
      <img
        alt="edX"
        className="d-block"
        src="https://edx-cdn.org/v3/default/logo.svg"
      />
    </a>
  </div>
  <div
    className="w-100 d-flex justify-content-end align-items-center"
  >
    <nav
      aria-label="Secondary"
      className="menu position-static"
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        aria-expanded={false}
        aria-haspopup="menu"
        aria-label="Account Menu"
        className="menu-trigger icon-button"
        onClick={[Function]}
        title="Account Menu"
      >
        <span
          className="avatar overflow-hidden d-inline-flex rounded-circle null"
          style={
            Object {
              "height": "1.5rem",
              "width": "1.5rem",
            }
          }
        >
          <svg
            aria-hidden={true}
            focusable="false"
            height="24px"
            role="img"
            style={
              Object {
                "height": "1.5rem",
                "width": "1.5rem",
              }
            }
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
          >
            <path
              d="M4.10255106,18.1351061 C4.7170266,16.0581859 8.01891846,14.4720277 12,14.4720277 C15.9810815,14.4720277 19.2829734,16.0581859 19.8974489,18.1351061 C21.215206,16.4412566 22,14.3122775 22,12 C22,6.4771525 17.5228475,2 12,2 C6.4771525,2 2,6.4771525 2,12 C2,14.3122775 2.78479405,16.4412566 4.10255106,18.1351061 Z M12,24 C5.372583,24 0,18.627417 0,12 C0,5.372583 5.372583,0 12,0 C18.627417,0 24,5.372583 24,12 C24,18.627417 18.627417,24 12,24 Z M12,13 C9.790861,13 8,11.209139 8,9 C8,6.790861 9.790861,5 12,5 C14.209139,5 16,6.790861 16,9 C16,11.209139 14.209139,13 12,13 Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </nav>
  </div>
</header>
`;
